### PR TITLE
Parser cleanup

### PR DESCRIPTION
**Pull request by kyouko-taiga**

### CODE REVIEW

File: Sources/Compiler/AST/Expr/LabeledArgument.swift [hunk 1]
```diff
@@ -1,13 +1,13 @@
-/// An argument of a function or subscript call.
-public struct CallArgument: Codable {
+/// A static argument or an argument to a function or subscript call.
+public struct LabeledArgument: Codable {
```
Comment (dabrahams) [inline Sources/Compiler/AST/Expr/LabeledArgument.swift:2]: Unless there are `UnlabeledArgument`s, I'm not sure this is an improved name.  Are we going to have `GenericArgument`s that are distinct?  If so, I like the old name.  Otherwise, I might opt for `Argument`.

File: Sources/Compiler/AST/Expr/LabeledArgument.swift [hunk 2]
```diff
@@ -1,13 +1,13 @@
...
 
   /// The label of the argument, if any.
   public let label: SourceRepresentable<Identifier>?
 
   /// The value of the argument.
   public let value: AnyExprID
 
-  public init(label: SourceRepresentable<Identifier>? = nil, value: AnyExprID) {
+  public init(label: SourceRepresentable<Identifier>?, value: AnyExprID) {
```
Comment (dabrahams) [inline Sources/Compiler/AST/Expr/LabeledArgument.swift:10]: Nice.

File: Sources/Compiler/Parse/Parser+Diagnostics.swift [hunk 3]
```diff
@@ -0,0 +1,59 @@
+extension Diagnostic {
+
+  static func diagnose(assignOperatorRequiresWhitespaces token: Token) -> Diagnostic {
+    .error("infix operator requires whitespaces on both sides", range: token.origin)
```
Comment (dabrahams) [inline Sources/Compiler/Parse/Parser+Diagnostics.swift:4]:
```suggestion
    .error("assignment operator requires whitespaces on both sides", range: token.origin)
```

File: Sources/Compiler/Parse/Parser+Diagnostics.swift [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  }
+
+  static func diagnose(expected kind: Token.Kind, at location: SourceLocation) -> Diagnostic {
+    .error("expected '\(kind)'", range: location ..< location)
```
Comment (dabrahams) [inline Sources/Compiler/Parse/Parser+Diagnostics.swift:8]: Shouldn't this one just dispatch to the next one?

File: Sources/Compiler/Parse/Parser+Diagnostics.swift [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      at: state.currentLocation,
+      children: [
+        .error(
+          "to match this '\(state.lexer.source[opener.origin])'",
+          range: opener.origin)
+      ]
+    )
+  }
+
+  static func diagnose(infixOperatorRequiresWhitespacesAt range: SourceRange?) -> Diagnostic {
```
Comment (dabrahams) [inline Sources/Compiler/Parse/Parser+Diagnostics.swift:35]: Can we stop calling these parameters `range` in new code at least?

File: Sources/Compiler/Parse/Parser+Diagnostics.swift [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+  static func diagnose(assignOperatorRequiresWhitespaces token: Token) -> Diagnostic {
+    .error("infix operator requires whitespaces on both sides", range: token.origin)
+  }
+
+  static func diagnose(expected kind: Token.Kind, at location: SourceLocation) -> Diagnostic {
+    .error("expected '\(kind)'", range: location ..< location)
+  }
+
+  static func diagnose(
```
Comment (dabrahams) [inline Sources/Compiler/Parse/Parser+Diagnostics.swift:11]: Not related to this PR, is the word `diagnose` really adding anything to any of these at the use site?  I kinda get the feeling that the first argument label should replace it.

File: Sources/Compiler/Parse/Parser+Diagnostics.swift [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+  static func diagnose(separatedPrefixOperatorAt range: SourceRange?) -> Diagnostic {
+    .error("prefix operator cannot be separated from its operand", range: range)
+  }
+
+  static func diagnose(unexpectedToken token: Token) -> Diagnostic {
+    .error("unexpected token '\(token.kind)'", range: token.origin)
+  }
+
+  static func diagnose(unterminatedCommentEndingAt endLocation: SourceLocation) -> Diagnostic {
```
Comment (dabrahams) [inline Sources/Compiler/Parse/Parser+Diagnostics.swift:51]: Seems like you probably want the start of these unterminated things as child diagnostics?

File: Sources/Compiler/Parse/Parser+Diagnostics.swift (hunk 6 above)
Comment (dabrahams) [reply Sources/Compiler/Parse/Parser+Diagnostics.swift:11]: Actually now that I look at the changes in `Parser.swift`, maybe it _is_ related to this PR.  Use sites are starting to look pretty complex and show signs of the Muffin Man.

File: Sources/Compiler/Parse/Parser.swift [hunk 8]
```diff
@@ -259,7 +259,7 @@ public enum Parser {
       if prologue.isEmpty {
         return nil
       } else {
-        throw DiagnosedError(expected("declaration", at: state.currentLocation))
+        throw DiagnosedError(.diagnose(expected: "declaration", at: state.currentLocation))
```
Comment (dabrahams) [inline Sources/Compiler/Parse/Parser.swift:262]: I don't see how the addition of the word `diagnose` is improving this use site.  Is there any reason these shouldn't all be `init`s of `DiagnosedError`?

File: Sources/Compiler/Parse/Parser.swift [hunk 9]
```diff
@@ -1329,11 +1291,6 @@ public enum Parser {
...
   )
 
-  static let parameterList = (
-    parameterDecl.and(zeroOrMany(take(.comma).and(parameterDecl).second))
-      .map({ (_, tree) -> [NodeID<ParameterDecl>] in [tree.0] + tree.1 })
-  )
-
   static let parameterDecl = (
     parameterInterface
       .and(maybe(take(.colon).and(parameterTypeExpr)))
```
Comment (dabrahams) [inline Sources/Compiler/Parse/Parser.swift:1296]: How did you decide which `.and`s to keep and which to drop?

File: Sources/Compiler/Parse/Parser.swift [hunk 10]
```diff
@@ -2198,33 +2097,28 @@ public enum Parser {
       wellFormed: SpawnExpr(decl: decl, origin: state.ast[decl].origin))
   }
 
-  private static func parseLambdaTypeOrTupleExpr(in state: inout ParserState) throws -> AnyExprID? {
+  private static func parseLambdaTypeOrTupleExpr(
+    in state: inout ParserState
+  ) throws -> AnyExprID? {
+    // Expect a left parenthesis.
```
Comment (dabrahams) [inline Sources/Compiler/Parse/Parser.swift:2103]: Is this new functionality?  Perhaps it should (in principle) be in a separate PR?

**Review by dabrahams — APPROVED** [review]
Very nice reduction!  Please see inline comments.

File: Sources/Compiler/AST/Expr/LabeledArgument.swift (hunk 1 above)
Comment (kyouko-taiga) [reply Sources/Compiler/AST/Expr/LabeledArgument.swift:2]: I wanted `Argument`, but it clashes with `swift-argument-parser` causes annoying issues with the property wrapper that we use in `CLI`.

I believe `GenericArgument` is gone with that PR.

**Comment by dabrahams** [comment]
> On Dec 9, 2022, at 6:01 PM, Dimi Racordon ***@***.***> wrote:
> 
> I wanted Argument, but it clashes with swift-argument-parser causes annoying issues with the property wrapper that we use in CLI.
> 
> 

Need `fileprivate import`!!

File: Sources/Compiler/Parse/Parser+Diagnostics.swift (hunk 4 above)
Comment (kyouko-taiga) [reply Sources/Compiler/Parse/Parser+Diagnostics.swift:8]: Makes sense

File: Sources/Compiler/Parse/Parser+Diagnostics.swift (hunk 5 above)
Comment (kyouko-taiga) [reply Sources/Compiler/Parse/Parser+Diagnostics.swift:35]: Sure, but most of these functions were just copy/pasted from a different file.

File: Sources/Compiler/Parse/Parser+Diagnostics.swift (hunk 6 above)
Comment (kyouko-taiga) [reply Sources/Compiler/Parse/Parser+Diagnostics.swift:11]: I used to have different function names, but that doesn't always work. Sometimes you really don't want a label for the first argument.

I'm open to suggestions but I prefer `diagnose(blah:blah:)` over `blah(_:blah:)`.

Comment (dabrahams) [reply Sources/Compiler/Parse/Parser+Diagnostics.swift:11]: Look at the use-sites.  Are you really writing what you'd want to read there?

File: Sources/Compiler/Parse/Parser+Diagnostics.swift (hunk 7 above)
Comment (kyouko-taiga) [reply Sources/Compiler/Parse/Parser+Diagnostics.swift:51]: That would make sense. Probably a good suggestion for an issue about improving diagnostics of unterminated comments.

File: Sources/Compiler/Parse/Parser.swift (hunk 8 above)
Comment (kyouko-taiga) [reply Sources/Compiler/Parse/Parser.swift:262]: That could be a solution yes. I don't mind `diagnose` but if you feel strongly against it, please open an issue so that we can do a PR that renames all uses.

File: Sources/Compiler/Parse/Parser.swift (hunk 9 above)
Comment (kyouko-taiga) [reply Sources/Compiler/Parse/Parser.swift:1296]: I want to remove all of them eventually.

File: Sources/Compiler/Parse/Parser+Diagnostics.swift (hunk 6 above)
Comment (kyouko-taiga) [reply Sources/Compiler/Parse/Parser+Diagnostics.swift:11]: I'm writing a lot of `diagnostics.append(.diagnose(blah: foo, blah: bar))` at call sites. Admittedly, that's little verbose, but I honestly think that's acceptable. As mentioned in another comment, I wouldn't mind using `diagnostics.append(.init(blah: foo, blah: bar))`.

Like I said, a previous version was designed do that you could write `diagnostics.append(.unexpected(token, at: currentLocation))`, which I guess you would prefer over the current approach. The problem was that it would not work well with things such as `diagnose(_:doesNotConformTo trait:at:because:)`(see `TypeChecker+Diagnostics.swift` for this example and others). I also felt it was starting to make `Diagnostic`'s namespace very messy. So ultimately I decided to keep the diagnostic constructors more uniform.

Comment (kyouko-taiga) [reply Sources/Compiler/Parse/Parser+Diagnostics.swift:11]: Excellent suggestion!
